### PR TITLE
Pipeline optimization (Sprint 34–36)

### PR DESCRIPTION
**Pull request by anthonyfok**
Dear all,

This PR represents interim result of the ongoing the add_data.sh pipeline optimization work as of ~Friday, May 21, 2021~ Friday, June 4, 2021 (status update on Monday, June 7).

- [x] Fetch from compressed repos (Stage 1):
    - [x] fetch_csv_xz function
    - [x] Move code to fetch_psra_csv_from_model and merge_csv functions
    - ~Fetch compressed CSV for PSRA and DSRA too: TODO~ Defer to future pull requests
- [x] Speed up file writes with eatmydata
- [x] Set synchronous_commit=off for speed
- [x] Add helper functions: LOG, RUN, INFO, WARN, ERROR
    - [x] Hide secrets (GITHUB_TOKEN, POSTGRES_PASS, ES_PASS) in LOG
    - [x] RUN(): Print line number and function too, with options to turn them off.
- [x] Add dry-run mode testing and debugging
- [x] Make add_data.sh and postgis/*.sh ShellCheck clean
- [x] Read variables from environment instead of command-line arguments (Fix potential curl error due to KIBANA_ENDPOINT being empty at "Creating PSRA Kibana Index Patterns")
- [x] Verification of file checksums (Stage 1)
    - [x] Fetch pointers of CSV files for "oid sha256"
    - ~To actually use these checksums~ Defer to a future pull request
- [x] Move all major steps into their own functions, and call them from a new main() function (ideas from Google Shell Style Guide).  To be used as prerequisite for #99

~It is a work-in-progress, and not quite ready for merging in light of the bugs/caveats below, but can be used for testing.  The simplest way to test is probably to check out the `pipeline-optimization` branch.~

2021-06-07 Update: I think it is finally ready for merging.  Pleasesee https://github.com/OpenDRR/opendrr-api/pull/105#issuecomment-856076511 for details.

---

### Caveats

* About 2/3 (?) of files are still being downloaded over Git LFS?
* I seem to have introduced at least the following new (non-aborting) error; yet to debug:
    ```
    ./add_data.sh: line 568: ((: --exposureAgg=b: attempted assignment to non-variable (error token is "=b")
    ```
    (Update: Not seen in my June 4-to-5 run.  The problem fixed itself?  Not sure...)
* As such, no guarantee of correctness (yet).
* ~~My last run crashed here, apparently due to a missing or empty KIBANA_ENDPOINT variable, but could be bug in my RUN() function too:~~ Resolved in commit 40e10fb9bdd3ca0326f24cc925191fba9e9f05c6 on June 3, 2021.
    ```
    2021-05-22T19:27:12.273954877Z Creating PSRA Kibana Index Patterns
    2021-05-22T19:27:12.304591179Z [add_data] curl -X POST -H 'securitytenant: global' -H 'Content-Type: application/json' /api/saved_objects/index-pattern/psra*all_indicators_s -H 'kbn-xsrf: true' -d '{ "attributes": { "title":"psra*all_indicators_s"}}'
    2021-05-22T19:27:12.314475601Z curl: (3) URL using bad/illegal format or missing URL
    2021-05-22T19:27:12.315562560Z Command exited with non-zero status 3
    ```
    Original code:
    ```bash
    echo "Creating PSRA Kibana Index Patterns"
    RUN curl -X POST -H "securitytenant: global" -H "Content-Type: application/json" "${KIBANA_ENDPOINT}/api/saved_objects/index-pattern/psra*all_indicators_s" -H "kbn-xsrf: true" -d '{ "attributes": { "title":"psra*all_indicators_s"}}'
    ```

### Observations

* An abnormality when fetching an LFS file in my last run: https://media.githubusercontent.com/media/OpenDRR/scenario-catalogue/master/FINISHED/s_shakemap_IDM7p1_Sidney_289.csv?token=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXX, a 4.4MB file, took over 22 minutes over a 100Mbps connection.  (#90)
* Compared to the newer PSRA Python scripts, the DSRA Python scripts (which Drew says predates the use of Docker instances) using are a lot more memory intensive, with the following taking 8.25 GB of RAM:
    ```
    [add_data] python3 DSRA_outputs2postgres_lfs.py --dsraModelDir=https://github.com/OpenDRR/scenario-catalogue/tree/master/FINISHED --columnsINI=DSRA_outputs2postgres.ini --eqScenario=SCM6p5_Ottawa
    126.51user 35.87system 7:25.93elapsed 36%CPU (0avgtext+0avgdata 8255000maxresident)k
    real     7m26.738s
    ```
    vs. the newer PSRA scripts apparently taking only 21 MB of RAM.  But need to verify more carefully the exact cause of the high memory usage in `OpenDRR/model-factory/DSRA_outputs2postgres_lfs.py`.  But yes, I am glad that Will suggested the need to refactor those older DSRA scripts (see OpenDRR/model-factory#53)

## Sample run log

See [2021-05-22-sample-run.log](https://github.com/OpenDRR/opendrr-api/files/6542748/2021-05-22-sample-run.log).

It was generated with:

    docker-compose logs -t --tail="all" > "logs/$(date -Is).log"
    egrep 'python-opendrr_1    |\[add_data\]|Z real|add_data\.sh' logs/2021-05-25T10:02:16-06:00.log | cut -b37- > 2021-05-22-sample-run.log

(TODO: quick-and-dirty commands; need to improve.)

### CODE REVIEW

**Comment by anthonyfok** [comment]
Hi @jvanulde and @drotheram,

I think this PR is finally ready for merging, so please kindly review at your convenience.

While all the origin goals are completed (please see the TODOs in the revised top post), my latest test run got up to "Load Social Fabric Views" until I ran out of memory.  As this is the 2nd last step, I am super lucky and happy!

```
python-opendrr_1  | 2021-06-05T16:19:07.537772692Z [add_data:809:import_data_from_postgis_to_elasticsearch] RUN: python3 socialFabric_postgres2es.py --type=all_indicators --aggregation=sauid --geometry=geom_poly --idField=Sauid
python-opendrr_1  | 2021-06-05T17:34:21.236350433Z Command terminated by signal 9
python-opendrr_1  | 2021-06-05T17:34:25.132522098Z 162.12user 33.91system 1:15:02elapsed 4%CPU (0avgtext+0avgdata 10087920maxresident)k
python-opendrr_1  | 2021-06-05T17:34:25.132546155Z 1361040inputs+0outputs (165614major+3378895minor)pagefaults 0swaps
```

I went a bit off-topic in this PR: Besides speed optimization, there are some not-directly-related bug fixes, and probably more on reorganization of the code.  In the future, I'll try to keep my PRs smaller and more focused.

Here are the logs for my recent run, separated by service:

* [2021-06-05_python-opendrr.log](https://github.com/OpenDRR/opendrr-api/files/6610450/2021-06-05_python-opendrr.log)
* [2021-06-05_db-opendrr.log](https://github.com/OpenDRR/opendrr-api/files/6610454/2021-06-05_db-opendrr.log)
* [2021-06-05_elasticsearch-opendrr.log](https://github.com/OpenDRR/opendrr-api/files/6610457/2021-06-05_elasticsearch-opendrr.log)
* [2021-06-05_kibana-opendrr.log](https://github.com/OpenDRR/opendrr-api/files/6610460/2021-06-05_kibana-opendrr.log)
* [2021-06-05_pygeoapi-opendrr.log](https://github.com/OpenDRR/opendrr-api/files/6610472/2021-06-05_pygeoapi-opendrr.log)

Alternatively, the combined full log: [2021-06-05_full.log](https://github.com/OpenDRR/opendrr-api/files/6610462/2021-06-05_full.log)

Thanks again!
(And again, my apologies for my delay in other tasks... will be getting to them soon ASAP!)

**Comment by drotheram** [comment]
Can we move the boundaries .dump file from backblaze to the boundaries repo instead? line 425-427 of add_data.sh

File: python/add_data.sh [hunk 1]
```diff
@@ -55,626 +167,723 @@ run_ogr2ogr() {
...
 
-run_psql Create_table_sovi_index_canada_v2.sql
-run_psql Create_table_sovi_census_canada.sql
-run_psql Create_table_sovi_thresholds.sql
+import_census_boundaries() {
+  LOG "## Importing Census Boundaries"
+
+  INFO "Trying to download pre-generated PostGIS database dump (for speed)..."
+  if RUN curl -O -v --retry 999 --retry-max-time 0 https://opendrr.eccp.ca/file/OpenDRR/opendrr-boundaries.dump || \
+    RUN curl -O -v --retry 999 --retry-max-time 0 https://f000.backblazeb2.com/file/OpenDRR/opendrr-boundaries.dump
```
Comment (drotheram) [inline python/add_data.sh:427]: Can we move this to the github repo or s3 bucket if needed?

File: python/add_data.sh [hunk 2]
```diff
@@ -55,626 +167,723 @@ run_ogr2ogr() {
...
+          GIT_LFS_SKIP_SMUDGE=1 git checkout )
+    done
+  )
+}
 
+# wait_for_postgres waits until PostGIS is ready to accept connections
+wait_for_postgres() {
+  LOG "Wait until PostgreSQL is ready"
+  until RUN pg_isready -h "$POSTGRES_HOST" -p "$POSTGRES_PORT" -U "$POSTGRES_USER"; do
+    sleep 2
```
Comment (drotheram) [inline python/add_data.sh:414]: $POSTGRES port is set to 5433 in .env file for the build process but the database still seems to only accept connections of 5432 (possibly related to the way the port range is defined in the docker-compose.yml, I'm not sure)

Comment (anthonyfok) [reply python/add_data.sh:414]: Hey Drew!  Thank you for catching that, and for your clear explanation.  I carelessly changed `-p 5432` to `-p "$POSTGRES_PORT"`, without documentation, while I was fixing ShellCheck warnings, in the following commit:

```
commit 313776fec815e194ee5c958e58aede796bce6bf2
Author: Anthony Fok <anthony.fok@canada.ca>
Date:   Wed Jun 2 17:30:35 2021 -0600

    Make add_data.sh and postgis/*.sh ShellCheck clean
    
    See PR #89
```

I had a hunch that I did something wrong, and thought of separating out those documented changes, but didn't...  My apologies!  I should definitely fix that somehow.  Thanks again for catching this!  I have introduced an incorrect fix, thus creating a bug here.  I'll follow-up on this soon.

File: python/add_data.sh (hunk 1 above)
Comment (anthonyfok) [reply python/add_data.sh:427]: Good point!  Will follow up at #116, probably in a follow-up PR.

**Comment by drotheram** [comment]
Since the code using backblaze2b has already been merged into the main branch, I'm proposing, if there are not other issues, that we merge into the main branch. We can treat the backblaze to s3 migration as a separate issue

**Comment by anthonyfok** [comment]
Thank you Drew for your amazing review, especially in how you have caught and resolved the bugs in the PR.

Let's "Rebase and merge" this in synchronous with OpenDRR/model-factory#66.
(I'll wait for you to review OpenDRR/model-factory#66 first.)

**Comment by drotheram** [comment]
Looks good for the most part so far but raw data ingest of PSRA results for YT into PostGIS is failing at my runtime. Need to investigate further. Otherwise everything else looks good!

File: python/add_data.sh [hunk 3]
```diff
@@ -11,40 +11,152 @@
...
-ES_ENDPOINT=$6
-ES_USER=$7
-ES_PASS=$8
-KIBANA_ENDPOINT=$9
+ADD_DATA_PRINT_FUNCNAME=${ADD_DATA_PRINT_FUNCNAME:-true}
+ADD_DATA_PRINT_LINENO=${ADD_DATA_PRINT_LINENO:-true}
 
 DSRA_REPOSITORY=https://github.com/OpenDRR/scenario-catalogue/tree/master/FINISHED
 
+PT_LIST=(AB BC MB NB NL NS NT NU ON PE QC SK)
```
Comment (drotheram) [inline python/add_data.sh:32]: YT (Yukon is missing from this list)

Comment (anthonyfok) [reply python/add_data.sh]: Thank you so much for catching this!
I introduced such a glaring brown-paper-bag bug... :sweat_smile:
Fixed in commit 29dec2c.

**Comment by anthonyfok** [comment]
Thanks a million Drew!  Total respect for your meticulous code review!

**Pull request merged**